# data acquisition / processing homework 1

> I pledge my Honor that I have abided by the Stevens Honor System. - Joshua Schmidt 2/12/21

In [53]:
import pandas as pd
import numpy as np

In [54]:
df = pd.read_csv('./data.csv')
print(df.columns)
unused_cols = ['Date']
df = df.drop(columns=unused_cols)
df

Index(['Date', 'Mkt-RF', 'SMB', 'HML', 'RF', 'Mom', 'Food', 'Beer', 'Smoke',
       'Games', 'Books', 'Hshld', 'Clths', 'Hlth', 'Chems', 'Txtls', 'Cnstr',
       'Steel', 'FabPr', 'ElcEq', 'Autos', 'Carry', 'Mines', 'Coal', 'Oil',
       'Util', 'Telcm', 'Servs', 'BusEq', 'Paper', 'Trans', 'Whlsl', 'Rtail',
       'Meals', 'Fin', 'Other'],
      dtype='object')


Mkt-RF   SMB   HML    RF   Mom  Food   Beer  Smoke  Games  Books  ...  \
0     -0.10 -0.09  4.72  0.25  0.36 -0.70   0.57  -0.33   2.46   9.67  ...   
1      4.32  0.31  3.40  0.26 -1.67  4.29  12.83   1.58   1.43   1.41  ...   
2      0.33 -1.77 -2.42  0.30  2.97  1.98 -13.56   5.55   0.57  -0.28  ...   
3      0.42  0.30  1.03  0.25  4.53  2.60   2.85   4.09  -3.34  -0.96  ...   
4      5.36  0.67  3.41  0.30  3.41  6.14  11.62  11.87  -0.50   3.13  ...   
..      ...   ...   ...   ...   ...   ...    ...    ...    ...    ...  ...   
943   -0.89 -0.94  1.40  0.30  2.22 -1.58   0.48   4.54  -0.86   1.61  ...   
944    0.77 -0.77  1.15  0.29  3.48  0.65   0.55   4.74  -1.77  -3.46  ...   
945   -2.35 -0.99 -0.78  0.27 -1.42 -3.46   0.58   1.70  -4.04  -6.26  ...   
946    3.73  1.01 -1.73  0.31  0.42 -0.02   1.47  -2.62   5.25   2.57  ...   
947    0.03 -0.48  0.47  0.32  0.75 -0.26  -2.14   4.21  -1.13  -1.65  ...   

     Telcm  Servs  BusEq  Paper  Trans  Whlsl  Rtail  Meals    Fin  Other  
0     1.88   2.08  -1.45  -2.60   1.44 -17.93  -3.34   1.53  -2.48  -4.13  
1     3.97   8.90   4.85   5.21   5.20   3.49   4.48   6.81   2.77   0.30  
2     5.56  -7.80   4.30  -8.39   1.06 -20.47   3.05  -2.44   1.41   2.28  
3    -2.08   3.44   3.10   4.43   0.77 -10.75   2.09   6.02   3.76   4.71  
4     3.35  18.33   5.10   5.66   6.69  -4.01   0.49   4.69  10.25   1.40  
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
943  -1.09   0.41  -1.53  -2.94  -3.46  -1.37  -6.50  -4.82  -1.44  -2.51  
944  -1.50  -0.02   1.40  -0.13   3.13  -0.01  -2.14  -0.29   1.14   0.61  
945  -3.23   1.63  -4.55   0.62   3.01  -1.87   2.29   0.16   2.49   0.20  
946   2.55   5.46   7.32   7.53   7.46   4.47   3.75   6.79   4.49   4.73  
947  -1.59  -2.12  -1.37   1.46   1.12   1.02  -1.09   1.03   0.25  -0.34  

[948 rows x 35 columns]

In [55]:
correlation_matrix = df.corr()
correlation_matrix

Mkt-RF       SMB       HML        RF       Mom      Food      Beer  \
Mkt-RF  1.000000  0.326863  0.216145 -0.068723 -0.338343  0.835924  0.707673   
SMB     0.326863  1.000000  0.094113 -0.059640 -0.164023  0.201698  0.351039   
HML     0.216145  0.094113  1.000000  0.012115 -0.400635  0.215132  0.214982   
RF     -0.068723 -0.059640  0.012115  1.000000  0.039130  0.032222 -0.011277   
Mom    -0.338343 -0.164023 -0.400635  0.039130  1.000000 -0.280289 -0.200077   
Food    0.835924  0.201698  0.215132  0.032222 -0.280289  1.000000  0.705154   
Beer    0.707673  0.351039  0.214982 -0.011277 -0.200077  0.705154  1.000000   
Smoke   0.584268  0.103154  0.171809  0.063036 -0.219165  0.670555  0.448606   
Games   0.830211  0.412089  0.250387 -0.024963 -0.356992  0.719049  0.677278   
Books   0.830092  0.408145  0.250608  0.009172 -0.306399  0.756656  0.622058   
Hshld   0.816234  0.261883  0.107373 -0.024209 -0.199683  0.788074  0.660953   
Clths   0.780630  0.460134  0.246719 -0.020186 -0.360482  0.714750  0.612181   
Hlth    0.804022  0.208896  0.064393  0.007783 -0.225778  0.784460  0.644274   
Chems   0.883889  0.208012  0.192392 -0.033856 -0.301203  0.749087  0.639947   
Txtls   0.823438  0.495356  0.347476 -0.035738 -0.371137  0.719605  0.683986   
Cnstr   0.913608  0.401464  0.247212 -0.031264 -0.323221  0.796304  0.718938   
Steel   0.867909  0.363120  0.343723 -0.048751 -0.394459  0.655520  0.568911   
FabPr   0.925046  0.406788  0.260264 -0.056682 -0.372352  0.738213  0.653430   
ElcEq   0.904103  0.257442  0.202407 -0.014688 -0.324623  0.760991  0.634189   
Autos   0.843297  0.285971  0.302826 -0.059864 -0.389308  0.681846  0.591363   
Carry   0.832370  0.362621  0.337407 -0.033650 -0.343814  0.707652  0.630898   
Mines   0.706160  0.343889  0.245741 -0.052210 -0.271214  0.554159  0.505533   
Coal    0.490542  0.267311  0.172241 -0.029589 -0.137088  0.376554  0.357831   
Oil     0.772597  0.137730  0.302340 -0.023956 -0.249883  0.620409  0.521132   
Util    0.761767  0.175149  0.354729  0.026696 -0.331665  0.714011  0.580965   
Telcm   0.755235  0.160554  0.099079  0.034355 -0.293874  0.651823  0.510663   
Servs   0.495606  0.285080 -0.084935 -0.011451 -0.092466  0.402551  0.346565   
BusEq   0.842644  0.329809 -0.066278 -0.051596 -0.228394  0.629850  0.542770   
Paper   0.869099  0.275653  0.197336 -0.024349 -0.300174  0.759879  0.633564   
Trans   0.858688  0.348099  0.430456 -0.017855 -0.423153  0.718959  0.620306   
Whlsl   0.796012  0.459826  0.209154 -0.012879 -0.260379  0.702854  0.688612   
Rtail   0.858747  0.302931  0.114997 -0.009125 -0.309388  0.812514  0.656186   
Meals   0.753859  0.368637  0.163557 -0.022962 -0.250813  0.716436  0.624511   
Fin     0.915884  0.282586  0.324566 -0.011648 -0.420421  0.822095  0.690496   
Other   0.842008  0.423908  0.165099 -0.035674 -0.290321  0.729381  0.637161   

           Smoke     Games     Books  ...     Telcm     Servs     BusEq  \
Mkt-RF  0.584268  0.830211  0.830092  ...  0.755235  0.495606  0.842644   
SMB     0.103154  0.412089  0.408145  ...  0.160554  0.285080  0.329809   
HML     0.171809  0.250387  0.250608  ...  0.099079 -0.084935 -0.066278   
RF      0.063036 -0.024963  0.009172  ...  0.034355 -0.011451 -0.051596   
Mom    -0.219165 -0.356992 -0.306399  ... -0.293874 -0.092466 -0.228394   
Food    0.670555  0.719049  0.756656  ...  0.651823  0.402551  0.629850   
Beer    0.448606  0.677278  0.622058  ...  0.510663  0.346565  0.542770   
Smoke   1.000000  0.506304  0.502045  ...  0.451984  0.287747  0.441467   
Games   0.506304  1.000000  0.738286  ...  0.618315  0.412909  0.717103   
Books   0.502045  0.738286  1.000000  ...  0.608323  0.476256  0.656846   
Hshld   0.543475  0.691324  0.725539  ...  0.594137  0.401277  0.683104   
Clths   0.493648  0.723338  0.731030  ...  0.549073  0.408214  0.633035   
Hlth    0.572477  0.692063  0.677021  ...  0.604775  0.400630  0.683920   
Chems   0.517167  0.716834  0.714475  ...  0.648008  0.381556  0.719747   
Txtl

In [56]:
factors = ['Mkt-RF', 'SMB', 'HML', 'RF', 'Mom']

max_factors, min_factors = [], []
used_cols = []
for col in df.columns:
    if col in factors:
        continue
    used_cols.append(col)
    max_factor, min_factor = None, None
    max_factor_val, min_factor_val = None, None
    for factor in factors:
        curr_val = correlation_matrix[col][factor]
        if max_factor is None or curr_val > max_factor_val:
            max_factor_val = curr_val
            max_factor = factor
        if min_factor is None or curr_val < min_factor_val:
            min_factor_val = curr_val
            min_factor = factor
    max_factors.append(max_factor)
    min_factors.append(min_factor)

factors_df = pd.DataFrame([max_factors, min_factors], columns=used_cols, index=['positive', 'negative'])
print('max & min correlation factors:', factors_df.to_numpy())

rf_data = correlation_matrix['RF'][[col for col in df.columns if col not in factors]]
# print(rf_data)
avg_rf_correlation = np.mean(rf_data)
print('average rf correlation:', avg_rf_correlation)

factors_df

max & min correlation factors: [['Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF'
  'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF'
  'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF'
  'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF' 'Mkt-RF']
 ['Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom'
  'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom'
  'Mom' 'Mom' 'Mom' 'Mom' 'Mom' 'Mom']]
average rf correlation: -0.01750517678862145


Food    Beer   Smoke   Games   Books   Hshld   Clths    Hlth  \
positive  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF   
negative     Mom     Mom     Mom     Mom     Mom     Mom     Mom     Mom   

           Chems   Txtls  ...   Telcm   Servs   BusEq   Paper   Trans   Whlsl  \
positive  Mkt-RF  Mkt-RF  ...  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF   
negative     Mom     Mom  ...     Mom     Mom     Mom     Mom     Mom     Mom   

           Rtail   Meals     Fin   Other  
positive  Mkt-RF  Mkt-RF  Mkt-RF  Mkt-RF  
negative     Mom     Mom     Mom     Mom  

[2 rows x 30 columns]

## part 1

Market-risk free correlates most highly with all of the provided industries. Momemntum correlates most negatively with every industry. With an average correlation for RF of -0.0175, RF does not correlate highly with the 30 industry time series

## part 2

Suppose that the daily value of a financial time series follows the model 

$X_t=0.01 + 0.2 X_{t-2} + a_t$

where $\{ a_t \}$ is a Gaussian white noise series with mean zero and variance 0.02.

- What are the mean and variance of the return series $X_t$?

### mean
  
$a_t$ non-deterministic, the rest deterministic  
  
$E\{X_t\} = E\{0.01 + 0.2 X_{t-2} + a_t\}$  
= $0.01 + E\{0.2 X_{t-2}\} + E\{a_t\}$  
= $0.01 + 0.2 E\{X_{t-2}\} + E\{a_t\}$  
$E\{a_t\} = 0$  
stationary => $E\{X_t\} = E\{X_{t-2}\}$  
$E\{a_t\} = 0.01 + 0.2 \cdot E\{a_t\} + 0$  
  
mean = E\{a_t\} = <mark style="background-color: lightblue">0.0125</mark>  
  
### variance

$\mu = E\{X_t\}$  
$X_t = 0.01 + 0.2 X_{t-2} + a_t$  
$(X_t - \mu) = 0.2 (X_{t - 2} - \mu) + a_t$  

$E\{X_t - \mu\}^2 = (0.2 (X_{t - 2} - \mu) + a_t)^2$  
$= E\{0.04 \cdot (X_{t - 2} - \mu)^2\} + E\{0.4 a_t \cdot (X_{t - 2} - \mu)\} + E\{a_t^2\}$  
$a_t$ is noise for today, independent from $X_{t-2}$. correlation is 0, second term = 0  
because it's stationary, $X_{t - 2} - \mu = variance$  
$var = 0.2^2 \cdot var + \sigma_a^2$, $\sigma_a^2$ is the noise level ($E\{a_t^2\}$)  

variance = <mark style="background-color: lightblue">$\frac{\sigma_a^2}{0.96}$</mark>  

- Compute the lag-1 and lag-2 autocorrelations of $X_t$. 

$X_t = 0.01 + 0.2 X_{t-2} + a_t$  
remove mean ($X_t = X_t - \mu$)  
$X_t - \mu = 0.01 + 0.2 X_{t-2} + a_t - \mu$  
$X_t - \mu = 0.2(X_{t-2} - \mu) + a_t$  
$\overline{X_t} = 0.2\overline{X}_{t-2} + a_t$  
  
auto-correlation:  
$E\{\overline{X_t} \cdot \overline{X}_{t-l}\} = E\{(0.2\overline{X}_{t-2} + a_t) \overline{X}_{t-l}\}$  
$E\{\overline{X_t} \cdot \overline{X}_{t-l}\} = E\{0.2 \cdot \overline{X}_{t-2} \cdot \overline{X}_{t-l}\}$  

lag 1:  
$E\{\overline{X_t} \cdot \overline{X}_{t-1}\} = E\{0.2 \cdot \overline{X}_{t-2} \cdot \overline{X}_{t-1}\}$  
lag 2:  
$E\{\overline{X_t} \cdot \overline{X}_{t-2}\} = E\{0.2 \cdot \overline{X}_{t-2} \cdot \overline{X}_{t-2}\}$  


- Assume that $X_{100} = -0.01$, and $X_{99} = 0.02$. Compute the 1- and 2-step-ahead forecasts of the return series at the forecast origin $t = 100$.

$\rho_l = 0.2 \rho_{l-2}$  
$\rho_{101} = 0.2 \rho_{99} = 0.2 \cdot 0.02 = 0.004$ (1-step-ahead forecast)  
$\rho_{102} = 0.2 \rho_{100} = 0.2 \cdot -0.01 = -0.002$ (2-step-ahead forecast)  
